source/modulo_core/src/Cell.cpp

```diff
@@ -419,6 +425,28 @@ bool Cell::on_cleanup() {
 
 rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn Cell::on_shutdown(const rclcpp_lifecycle::State& state) {
   RCUTILS_LOG_INFO_NAMED(get_name(), "on_shutdown() is called from state %s.", state.label().c_str());
+  uint8_t current_state = state.id();
+  // if the node is already shutdown just return success
+  if (state.id() == lifecycle_msgs::msg::State::PRIMARY_STATE_FINALIZED) {
+    return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS;
+  }
+  // check current state and eventually deactivate and deconfigure
+  if (current_state == lifecycle_msgs::msg::State::PRIMARY_STATE_ACTIVE) {
+    RCUTILS_LOG_INFO_NAMED(get_name(), "Node is active, deactivating it before shutdown.");
+    auto callback_code = this->on_deactivate(this->get_current_state());
+    if (callback_code == rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::FAILURE) {
+      return callback_code;
+    }
```

Comment:
I would again suggest comparing to "not success", and the syntax can be condensed.
```suggestion
    if (auto callback_return = this->on_deactivate(this->get_current_state()) != CallbackReturn::SUCCESS) {
      return callback_return;
    }
```